wordpress-mobile/WordPress-iOS
# Integrate PHPickerViewController in Page Editor

Part of https://github.com/wordpress-mobile/WordPress-iOS/issues/21190.

The changes affect both the WP and the Jetpack app.

https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/e9b6e5d9-2b60-43c5-88b8-db08220cbb4a

## Tests

> [!IMPORTANT]
> - Enable the "Native Photo Picker" feature flag before testing

- Verify that the Image block uploads work
- Verify that the Gallery block uploads work (make sure the order of inserted photos matches the selection)
- Verify that the Video block uploads work

For additional test cases, see https://github.com/wordpress-mobile/WordPress-iOS/pull/21343. It covers the picker and the media uploads at great lengths. But there should be no need to repeat them because this PR builds on top of these changes.

## [Tech] Preview URL

When the media is added to one of the Gutenberg blocks, it expects a URL. But the app has no URLs at the time the user makes a selection in the native picker.

In the current [implementation](https://github.com/wordpress-mobile/WordPress-iOS/blob/74bd0867f963a0fad2a7267f42f85034730b334a/WordPress/Classes/ViewRelated/Gutenberg/GutenbergMediaInserterHelper.swift#L79), when you select the assets, the editor first requests their previews using `PHImageManager.default().requestImage` and `.fastFormat` and only then updates the Gutenberg blocks. It achieves that when the blocks are updated, the images already have previews, but it comes at a great cost.

The requested thumbnails are usually very small: [example](https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/08a9cc47-9036-4a0e-89ca-beef420b3604) (120 ×90px), but with the next line `thumbImage.resizedImage(asset.pixelSize()`, the editor blows it up to the original image size: [example](https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/3feeff61-ee63-41b4-be64-9d6bdedc5681) (4032 × 3024px) and puts them on disk in the temporary folder. There are many problems with it:

- The requests are performed in parallel with the work `MediaImportService` does to export the images and generate the previews (work duplication)
- There is no limit on the number of concurrent requests
- The resizing is expensive, and both the CPU and the RAM usage spike when you add the images: 480 MB → 680 MB when adding just five images
- If the assets in the cloud, the blocks won't get updated until they are downloaded

As a result, the app ends up wasting a ton of CPU, RAM, and disk space just to show previews that almost immediately (a few hundred milliseconds later) get replaced by `MediaImportService` in the following callback:

```swift
 private func mediaObserver(media: Media, state: MediaCoordinator.MediaState) {
        let mediaUploadID = media.gutenbergUploadID
        case .thumbnailReady(let url):
            guard ReachabilityUtils.isInternetReachable() && media.remoteStatus != .failed else {
                gutenberg.mediaUploadUpdate(id: mediaUploadID, state: .failed, progress: 0, url: url, serverID: nil)
                return
            }
            gutenberg.mediaUploadUpdate(id: mediaUploadID, state: .uploading, progress: 0.20, url: url, serverID: nil)
            // ...
```

The solution I went with in this PR was to **not** generate the thumbnails at all and just pass placeholder URLs that point to nothing. It dramatically reduces the memory footprint and there is no delay after the selection is made in the picker:

- Before: 480 MB → 680 MB (peak) (+200 MB)
- After: 490 MB → 505 MB (peak) (+15 MB)

The previews get displayed as soon as the `MediaImportService` generates them.

## Regression Notes
1. Potential unintended areas of impact: Media uploads in Post and Page Editor
2. What I did to test those areas of impact (or what existing automated tests I relied on): manual tests
3. What automated tests I added (or what prevented me from doing so): n/a

PR submission checklist:


UI Changes testing checklist:
